File: 2020/day12/main.cpp
#include <cassert>
#include <fstream>
#include <iostream>
#include <vector>

using std::cout;
using std::endl;
using std::ifstream;
using std::string;
using std::vector;

void puzzle1(vector<string> instructions) {
    int de = 1, dn = 0;
    int e = 0, n = 0;
    for (const auto& instruction: instructions) {
        char code = instruction[0];
        int value = stoi(instruction.substr(1));
        switch (code) {
        case 'E': e += value; break;
        case 'W': e -= value; break;
        case 'N': n += value; break;
        case 'S': n -= value; break;
        case 'R':
            value = 360 - value;
        case 'L':
            assert(value % 90 == 0);
            value %= 360;
            while (value) {
                int tmp = de;
                de = -dn;
                dn = tmp;
                value -= 90;
            }
            break;
        case 'F':
            n += dn * value;
            e += de * value;
            break;
        default:
            assert(0);
        }
    }
    cout << abs(e) + abs(n) << endl;
}

void puzzle2(vector<string> instructions) {
    int se = 0, sn = 0;
    int we = 10, wn = 1;
    for (const auto& instruction: instructions) {
        char code = instruction[0];
        int value = stoi(instruction.substr(1));
        switch (code) {
        case 'E': we += value; break;
        case 'W': we -= value; break;
        case 'N': wn += value; break;
        case 'S': wn -= value; break;
        case 'L':
        case 'R':
            if (code == 'R') {
                value = 360 - value;
            }
            assert(value % 90 == 0);
            value %= 360;
            while (value) {
                int tmp = we;
                we = -wn;
                wn = tmp;
                value -= 90;
            }
            break;
        case 'F':
            sn += wn * value;
            se += we * value;
            break;
        default:
            assert(0);
        }
    }
    cout << abs(se) + abs(sn) << endl;
}

int main(void) {
    ifstream f("input");
    assert(f.is_open());
    vector<string> instructions;
    string line;
    while (getline(f, line)) {
        instructions.push_back(line);
    }

    puzzle1(instructions);
    puzzle2(instructions);
}

File: 2020/day12/main.py
from typing import List


def puzzle1(instructions: List[str]) -> None:
    de, dn = 1, 0
    e, n = 0, 0
    for instruction in instructions:
        code = instruction[0]
        value = int(instruction[1:])
        assert value >= 0
        if code == 'E':
            e += value
        elif code == 'W':
            e -= value
        elif code == 'N':
            n += value
        elif code == 'S':
            n -= value
        elif code in ('L', 'R'):
            assert value % 90 == 0
            if code == 'R':
                value = 360 - value
            while value:
                de, dn = -dn, de
                value -= 90
        elif code == 'F':
            n += dn * value
            e += de * value
    print(abs(n) + abs(e))


def puzzle2(instructions: List[str]) -> None:
    se, sn = 0, 0
    we, wn = 10, 1
    for instruction in instructions:
        code = instruction[0]
        value = int(instruction[1:])
        assert value >= 0
        if code == 'E':
            we += value
        elif code == 'W':
            we -= value
        elif code == 'N':
            wn += value
        elif code == 'S':
            wn -= value
        elif code in ('L', 'R'):
            assert value % 90 == 0
            if code == 'R':
                value = 360 - value
            while value:
                we, wn = -wn, we
                value -= 90
        elif code == 'F':
            sn += wn * value
            se += we * value
    print(abs(sn) + abs(se))


def main() -> None:
    with open('input') as f:
        instructions = f.readlines()

    puzzle1(instructions)
    puzzle2(instructions)


if __name__ == '__main__':
    main()
